src/fsharp/FSharp.Core/reflect.fs

```diff
@@ -404,8 +404,10 @@ module internal Impl =
                 | _ -> invalidArg "tys" (SR.GetString(SR.invalidTupleTypes))
 
             let tables = if isStruct then valueTupleTypes else refTupleTypes
-            match tables.TryGetValue(asm) with
+            match lock dictionaryLock (fun () -> tables.TryGetValue(asm)) with
```

Comment:
I don't think this lock is necessary.  We check tables outside of a lock to see if it contains asm If it is not in tables then we take a lock, recheck the dictionary and if it hasn't been added since we first checked, then we add it in.  So as best I can tell there isn't really a race.  And we have added an unnecessary lock per read.

Does that make sense?

Kevin